Feat: Added a top-level Error to the App

Murdock9803 opened the pull request:
Closes: #637 

### Description
The pull request aims to add a top level Error Boundary to the app, so that any errors are shown gracefully on the screen. This will prevent a White screen in case of errors in rendering, and will present a message screen with `Something went wrong` message.

### Screenshots
![Screenshot 2025-03-19 084626](https://github.com/user-attachments/assets/0ca5f75d-8eca-4cd7-b6bb-5df4fffa34de)

### Checklist
- [x] Written clean code to catch errors and present an error screen.
- [x] Error will be shown at the console.
- [x] Tested the solution to work locally.
- [x] Tried to keep the code-style consistent with the other files.

alexlwn123 commented:
Hi @Murdock9803! Welcome & thanks for the contribution!

>  I was confused about what webSDK URL (wss://) to enter in the Homepage

Just to clarify, this is the [websocket](https://en.wikipedia.org/wiki/WebSocket) endpoint for the guardian server. The Fedimint Web SDK is something else

Will provide review in the PR

alexlwn123 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,92 @@
+            Something went wrong
+          </Text>
+
+          <a href='/'>
```
the `a` tag is fine for the top-level route

Murdock9803 replied:
I have included the `<Link>` as now our Error Boundary is at all the routes, and everything works fine. Please suggest if we should change it back to `a` 😄 🙌

alexlwn123 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,92 @@
+          </Text>
+
+          <a href='/'>
+            <Button colorScheme='blue'>Return to Home Page</Button>
```
can we add an additional button to reload the current page? This will only be possible after moving the ErrorBoundary within the Route.

Murdock9803 replied:
I have added the Reload button 🙌 The width for both the buttons is kept the same, for uniformity in design 🙌 
thanks for the review 🙌

alexlwn123 reviewed (changes requested):
overall looks good. Left some comments to address before merging. 

I suggest trying to closely follow the existing patterns in the codebase already. Typically, if we're already doing something in the codebase, we'll want to do it the same way everywhere

Murdock9803 commented on apps/router/src/App.tsx:
```diff
@@ -6,33 +6,36 @@ import { Guardian } from './guardian-ui/Guardian';
+          />
+        </Routes>
+      </Router>
+    </ErrorBoundary>
```
@alexlwn123 I have put the ErrorBoundary in the `Wrapper.tsx` component, and it makes sure the error boundary is present in all the routes 🙌 Please suggest any further adjustments, if any 🙌

Murdock9803 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,92 @@
+            <path
+              fillRule='evenodd'
+              clipRule='evenodd'
+              d='M26.8363 3.68341C26.8363 5.71771 25.1872 7.36683 23.1529 7.36683C21.1186 7.36683 19.4695 5.71771 19.4695 3.68341C19.4695 1.64912 21.1186 0 23.1529 0C25.1872 0 26.8363 1.64912 26.8363 3.68341ZM42.4491 11.205C42.4491 13.2393 40.8 14.8885 38.7657 14.8885C36.7314 14.8885 35.0822 13.2393 35.0822 11.205C35.0822 9.17074 36.7314 7.52163 38.7657 7.52163C40.8 7.52163 42.4491 9.17074 42.449 ...
```
@alexlwn123 I speculated this could be a problem, but since the error boundary was out of `Routes`, I couldn't include the <Logo /> SVG Component. The problem is solved now. Thanks for this 🙌

Murdock9803 commented:
Hello @alexlwn123 
I have made changes according to the reviews, and I'd be happy to make further changes, if required, to make the PR perfect 🙌 

<br/>
 
> Just to clarify, this is the [websocket](https://en.wikipedia.org/wiki/WebSocket) endpoint for the guardian server. The Fedimint Web SDK is something else.

Thanks for this, I will definitely read more about this and try to learn more about Fedimint too 🙌 My plan for few upcoming days is to read more and more about Fedimint and Bitcoin 🙌 

<br/>

> I suggest trying to closely follow the existing patterns in the codebase already. Typically, if we're already doing something in the codebase, we'll want to do it the same way everywhere.

Actually this is what I also aim for, but got confused with the `Routes` thing in this one. I will certainly remember to keep the code style consistent from now onwards 🙌 
Thanks for these reviews, I really appreciate it ☺️ 🙌

tsmith123 commented:
Could you please add some notes on how you've tested this and also please update the initial screenshot now that it has changed 🙏

Murdock9803 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,76 @@
+  Reload = () => {
+    this.setState({ hasError: false });
+    window.location.reload();
+  };
```
Actually I did not assign `const` with the function name, as it is a member of the Class. So it gave error.
Also I have removed the function entirely, as you suggested. `hasError` setting to `false` was really not needed, I should have looked upon it.. I'll try to be more careful next time 🙌

tsmith123 replied:
Yeah ignore that suggestion about the `const` - I forgot that this was a class component. Don't see many of them around these days. 😆

Murdock9803 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,76 @@
+          <Button
+            width={52}
+            marginTop={2}
+            onClick={this.Reload}
```
I used `windows.location.reload` inside the onClick directly, but we can't use `useNavigate` hook as it is a class component (`ErrorBoundary`). Should I research more about what can be done ?

Murdock9803 commented:
@tsmith123 Thanks for the review 🚀 
I have updated the screenshot after the recent changes, the current state of the Error Boundary screen.
I have also tried to implement your suggestions to make the code better 🙌 

> Could you please add some notes on how you've tested this

I tested this by implementing the following two steps:
1. bypassing the homepage screen.
2. creating an intentional error in `guardian UI` page.

To bypass the homepage, I changed `App.tsx` from `<Route path='/' element={<HomePage />} />` to ` <Route path="/" element={<Navigate to="/guardians/1" />} />`

To create an intentional error in `Guardian.tsx` : ` throw new Error("Error in Guardian component");`

This is how I tested the functioning and UI of the Error Boundary screen 🙌 Once again, thanks 🙌

tsmith123 commented:
Thanks @Murdock9803 - will pull this down and give it a test myself when I get a chance.

tsmith123 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,71 @@
+            color='red.500'
+          >
+            Something went wrong!
+          </Text>
```
This would need a translation. Don't worry about doing the translations but you just need to add the en version and hook it up to use this string.

Murdock9803 replied:
Thanks for pointing this out 😄  I have added the translations 🙌

tsmith123 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,71 @@
+            onClick={() => window.location.reload()}
+            colorScheme='blue'
+          >
+            Reload page
```
Translation

Murdock9803 replied:
Done 🙌

tsmith123 commented on apps/router/src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,71 @@
+
+          <Link to={'/'}>
+            <Button width={52} colorScheme='blue'>
+              Return to Home Page
```
Translation

Murdock9803 replied:
Done 🙌

Murdock9803 commented:
@tsmith123 
I have added the translations. At first, I thought it was straightforward but as the `ErrorBoundary` is a class component, it became somewhat different. I got to learn a new thing and the texts work fine in the screen 🙌
